File: vbxe_converter.js
#!/usr/bin/env node
/**
 * VBXE Image Converter for Atari 8-bit Browser
 * 
 * Converts web images to VBXE-compatible format:
 * - Header: [width_lo] [width_hi] [height]  (3 bytes)
 * - Palette: 256 colors × 3 RGB bytes (768 bytes)
 * - Pixels: raw 8-bit indexed (w × h bytes)
 * 
 * Usage:
 *   node vbxe_converter.js <image_url> [width] [height]
 * 
 * Example:
 *   node vbxe_converter.js https://example.com/image.jpg
 *   node vbxe_converter.js https://example.com/image.jpg 160 100
 */

const https = require('https');
const http = require('http');
const { Image } = require('image-js');

// VBXE palette indices 0-7 are reserved for text
const PALETTE_START = 8;

async function fetchImage(url) {
    return new Promise((resolve, reject) => {
        const protocol = url.startsWith('https') ? https : http;
        const request = protocol.get(url, {
            headers: {
                'User-Agent': 'VBXE Browser/1.0 Atari-Image-Converter'
            },
            timeout: 30000
        }, (response) => {
            if (response.statusCode !== 200) {
                reject(new Error(`HTTP ${response.statusCode}`));
                return;
            }
            
            const chunks = [];
            response.on('data', chunk => chunks.push(chunk));
            response.on('end', () => {
                const buffer = Buffer.concat(chunks);
                Image.load(buffer).then(resolve).catch(reject);
            });
            response.on('error', reject);
        });
        request.on('error', reject);
        request.on('timeout', () => {
            request.destroy();
            reject(new Error('Request timeout'));
        });
    });
}

function resizeImage(img, maxW, maxH) {
    const scaleW = maxW / img.width;
    const scaleH = maxH / img.height;
    const scale = Math.min(scaleW, scaleH, 1.0);
    
    const newW = Math.max(1, Math.round(img.width * scale));
    const newH = Math.max(1, Math.round(img.height * scale));
    
    // Create centered black canvas
    const canvas = new Image({
        width: maxW,
        height: maxH,
        color: [0, 0, 0, 255]
    });
    
    const resized = img.resize({ width: newW, height: newH });
    
    const xOffset = Math.floor((maxW - newW) / 2);
    const yOffset = Math.floor((maxH - newH) / 2);
    
    // Copy pixels to centered position
    for (let y = 0; y < newH; y++) {
        for (let x = 0; x < newW; x++) {
            const srcIdx = (y * newW + x) * 4;
            const dstIdx = ((yOffset + y) * maxW + (xOffset + x)) * 4;
            canvas.data[dstIdx] = resized.data[srcIdx];
            canvas.data[dstIdx + 1] = resized.data[srcIdx + 1];
            canvas.data[dstIdx + 2] = resized.data[srcIdx + 2];
            canvas.data[dstIdx + 3] = 255;
        }
    }
    
    return canvas;
}

// Simple median cut color quantization
function medianCut(pixels, depth) {
    if (depth === 0 || pixels.length <= 1) {
        const count = pixels.length || 1;
        let r = 0, g = 0, b = 0;
        for (const p of pixels) {
            r += p[0];
            g += p[1];
            b += p[2];
        }
        return [[Math.round(r / count), Math.round(g / count), Math.round(b / count)]];
    }
    
    // Find channel with largest range
    const rVals = pixels.map(p => p[0]);
    const gVals = pixels.map(p => p[1]);
    const bVals = pixels.map(p => p[2]);
    
    const rRange = Math.max(...rVals) - Math.min(...rVals);
    const gRange = Math.max(...gVals) - Math.min(...gVals);
    const bRange = Math.max(...bVals) - Math.min(...bVals);
    
    let sortKey;
    if (rRange >= gRange && rRange >= bRange) sortKey = p => p[0];
    else if (gRange >= bRange) sortKey = p => p[1];
    else sortKey = p => p[2];
    
    pixels.sort((a, b) => sortKey(a) - sortKey(b));
    
    const mid = Math.floor(pixels.length / 2);
    return [
        ...medianCut(pixels.slice(0, mid), depth - 1),
        ...medianCut(pixels.slice(mid), depth - 1)
    ];
}

function quantizeTo256(img) {
    const width = img.width;
    const height = img.height;
    
    // Extract RGB pixels
    const pixels = [];
    for (let i = 0; i < width * height; i++) {
        const r = img.data[i * 4];
        const g = img.data[i * 4 + 1];
        const b = img.data[i * 4 + 2];
        pixels.push([r, g, b]);
    }
    
    // Generate 248 colors (indices 8-255)
    let colors = medianCut(pixels, 8);
    colors = colors.slice(0, 248);
    
    // Pad to 248 if needed
    while (colors.length < 248) {
        colors.push([0, 0, 0]);
    }
    
    // Color distance for nearest match
    const colorDist = (c1, c2) => 
        Math.sqrt((c1[0] - c2[0]) ** 2 + (c1[1] - c2[1]) ** 2 + (c1[2] - c2[2]) ** 2);
    
    // Build pixel index data
    const indexData = [];
    const colorMap = new Map();
    
    for (const pixel of pixels) {
        const key = pixel.join(',');
        if (!colorMap.has(key)) {
            let minDist = Infinity;
            let bestIdx = 0;
            for (let idx = 0; idx < colors.length; idx++) {
                const dist = colorDist(pixel, colors[idx]);
                if (dist < minDist) {
                    minDist = dist;
                    bestIdx = idx;
                }
            }
            colorMap.set(key, bestIdx);
        }
        indexData.push(colorMap.get(key));
    }
    
    // Build palette as RGB bytes
    const paletteData = Buffer.alloc(768);
    for (let i = 0; i < 248; i++) {
        paletteData[i * 3] = colors[i][0];
        paletteData[i * 3 + 1] = colors[i][1];
        paletteData[i * 3 + 2] = colors[i][2];
    }
    
    return { paletteData, indexData, width, height };
}

function outputVBXE(width, height, palette, pixels) {
    const parts = [];
    
    // Header: width_lo, width_hi, height
    parts.push(Buffer.from([width & 0xFF, (width >> 8) & 0xFF, height & 0xFF]));
    
    // Palette: 8 reserved + 248 image colors = 256 colors = 768 bytes
    // Reserved colors (0-7) are black
    const reserved = Buffer.alloc(24); // 8 colors × 3 bytes
    parts.push(reserved);
    parts.push(palette);
    
    // Pixel data
    parts.push(Buffer.from(pixels));
    
    // Output to stdout
    process.stdout.write(Buffer.concat(parts));
}

async function main() {
    const args = process.argv.slice(2);
    if (args.length < 1) {
        console.error('Usage: node vbxe_converter.js <image_url> [width] [height]');
        process.exit(1);
    }
    
    const imageUrl = args[0];
    let maxWidth = parseInt(args[1]) || 320;
    let maxHeight = parseInt(args[2]) || 208;
    
    maxWidth = Math.max(8, Math.min(320, maxWidth));
    maxHeight = Math.max(8, Math.min(208, maxHeight));
    
    try {
        const img = await fetchImage(imageUrl);
        const resized = resizeImage(img, maxWidth, maxHeight);
        const { paletteData, indexData, width, height } = quantizeTo256(resized);
        outputVBXE(width, height, paletteData, indexData);
    } catch (error) {
        console.error(`Error: ${error.message}`);
        process.exit(1);
    }
}

main();

File: vbxe_converter.py
#!/usr/bin/env python3
"""
VBXE Image Converter for Atari 8-bit Browser

Converts web images to VBXE-compatible format:
- Header: [width_lo] [width_hi] [height]  (3 bytes)
- Palette: 256 colors × 3 RGB bytes (768 bytes)
- Pixels: raw 8-bit indexed (w × h bytes)

Usage:
    python vbxe_converter.py <image_url> [width] [height]
    
    image_url: URL of the image to convert
    width: max output width (default: 320)
    height: max output height (default: 208)
    
Example:
    python vbxe_converter.py https://example.com/image.jpg
    python vbxe_converter.py https://example.com/image.jpg 160 100
"""

import sys
import io
import math
from PIL import Image
import requests

# VBXE palette indices 0-7 are reserved for text, so we use 8-255 for image
PALETTE_START = 8


def fetch_image(url: str) -> Image.Image:
    """Download and return PIL Image from URL."""
    headers = {
        'User-Agent': 'VBXE Browser/1.0 Atari-Image-Converter'
    }
    response = requests.get(url, headers=headers, timeout=30)
    response.raise_for_status()
    return Image.open(io.BytesIO(response.content)).convert('RGBA')


def resize_image(img: Image.Image, max_w: int, max_h: int) -> Image.Image:
    """Resize image to fit within max dimensions, centered on black background."""
    # Calculate scale to fit
    scale_w = max_w / img.width
    scale_h = max_h / img.height
    scale = min(scale_w, scale_h, 1.0)  # Don't upscale
    
    new_w = max(1, int(img.width * scale))
    new_h = max(1, int(img.height * scale))
    
    # Resize with high quality
    resized = img.resize((new_w, new_h), Image.Resampling.LANCZOS)
    
    # Center on black background
    result = Image.new('RGBA', (max_w, max_h), (0, 0, 0, 255))
    x_offset = (max_w - new_w) // 2
    y_offset = (max_h - new_h) // 2
    result.paste(resized, (x_offset, y_offset))
    
    return result


def quantize_to_256(img: Image.Image) -> tuple:
    """
    Quantize image to 256 colors using median cut.
    Returns: (palette_data, pixel_data, width, height)
    """
    # Convert to RGB for processing
    rgb_img = img.convert('RGB')
    pixels = list(rgb_img.getdata())
    
    width, height = img.size
    
    # Median cut color quantization
    def median_cut(pixels_list, depth):
        if depth == 0 or len(pixels_list) <= 1:
            # Calculate average color
            r_sum = g_sum = b_sum = 0
            for r, g, b in pixels_list:
                r_sum += r
                g_sum += g
                b_sum += b
            count = len(pixels_list) or 1
            return [(r_sum // count, g_sum // count, b_sum // count)]
        
        # Find the channel with the largest range
        r_vals = [p[0] for p in pixels_list]
        g_vals = [p[1] for p in pixels_list]
        b_vals = [p[2] for p in pixels_list]
        
        r_range = max(r_vals) - min(r_vals)
        g_range = max(g_vals) - min(g_vals)
        b_range = max(b_vals) - min(b_vals)
        
        if r_range >= g_range and r_range >= b_range:
            pixels_list.sort(key=lambda p: p[0])
        elif g_range >= b_range:
            pixels_list.sort(key=lambda p: p[1])
        else:
            pixels_list.sort(key=lambda p: p[2])
        
        mid = len(pixels_list) // 2
        left = median_cut(pixels_list[:mid], depth - 1)
        right = median_cut(pixels_list[mid:], depth - 1)
        return left + right
    
    # Generate 248 colors (8 reserved for text)
    colors = median_cut(pixels, 8)  # 256 colors, but reserve 8
    colors = colors[:248]  # Ensure we have exactly 248
    
    # Pad to 248 if needed
    while len(colors) < 248:
        colors.append((0, 0, 0))
    
    # Build color lookup
    color_map = {}
    def color_distance(c1, c2):
        return math.sqrt(sum((a - b) ** 2 for a, b in zip(c1, c2)))
    
    # Create indexed pixel data
    index_data = []
    for pixel in pixels:
        # Find closest color in palette
        if pixel not in color_map:
            min_dist = float('inf')
            best_idx = 0
            for idx, c in enumerate(colors):
                dist = color_distance(pixel, c)
                if dist < min_dist:
                    min_dist = dist
                    best_idx = idx
            color_map[pixel] = best_idx
        index_data.append(color_map[pixel])
    
    # Convert palette to RGB bytes
    palette_data = bytearray()
    for r, g, b in colors:
        palette_data.append(r)
        palette_data.append(g)
        palette_data.append(b)
    
    return palette_data, index_data, width, height


def output_vbxe_format(width: int, height: int, palette: bytes, pixels: list):
    """Output the VBXE binary format to stdout."""
    # Header: width_lo, width_hi, height
    sys.stdout.buffer.write(bytes([width & 0xFF, (width >> 8) & 0xFF, height & 0xFF]))
    
    # Palette: 768 bytes (256 colors × 3)
    # First 8 colors (indices 0-7) are reserved, we fill with black
    reserved = bytes(24)  # 8 colors × 3 bytes
    sys.stdout.buffer.write(reserved)
    sys.stdout.buffer.write(palette)
    
    # Pad palette to 256 colors (768 bytes total)
    # We already wrote 24 bytes + 744 bytes = 768, but need to ensure exactly 768
    # Actually our palette is 248 colors = 744 bytes, plus 24 reserved = 768
    # But we need to write exactly 768 bytes for palette
    # The above should be correct: 8 reserved + 248 image colors = 256 total
    
    # Pixels: raw 8-bit indexed
    sys.stdout.buffer.write(bytes(pixels))


def main():
    if len(sys.argv) < 2:
        print(__doc__)
        sys.exit(1)
    
    image_url = sys.argv[1]
    max_width = int(sys.argv[2]) if len(sys.argv) > 2 else 320
    max_height = int(sys.argv[3]) if len(sys.argv) > 3 else 208
    
    # Clamp dimensions
    max_width = max(8, min(320, max_width))
    max_height = max(8, min(208, max_height))
    
    try:
        # Fetch and process image
        img = fetch_image(image_url)
        img = resize_image(img, max_width, max_height)
        palette, pixels, width, height = quantize_to_256(img)
        
        # Output VBXE format
        output_vbxe_format(width, height, bytes(palette), pixels)
        
    except requests.RequestException as e:
        sys.stderr.write(f"Error fetching image: {e}\n")
        sys.exit(1)
    except Exception as e:
        sys.stderr.write(f"Error processing image: {e}\n")
        sys.exit(1)


if __name__ == '__main__':
    main()